Misc fixes, including correctness

BryceStevenWilley opened the pull request:
* Correctly calculate poverty levels: see https://github.com/SuffolkLITLab/docassemble-MAPovertyScale/issues/5
* change a yes no button to a radio button, as we recommend for other forms
* correct a problem where setting `reconsider: True` to blocks would result in blanks in the PDF
  This happened because the value of the block would be invalidated on each load, however upon encountering the
  undefined value, the attachment with `skip undefined: True` would simply output a blank, and not reevaluate the value. This seems to only happen when the page is refreshed when generating the PDF, but could be recreated elsewhere likely.
* adds a `;` between the case name and number, which could cause readability issues for clerks
* if the user says they don't know the docket number, makes the follow up question not required. It is unclear what they should put if it's a new case, and the help even says to leave it blank, which isn't possible if the field is required
* shows the poverty level for the household size when the user says they can't pay but aren't under the 125% of poverty level
* adds a `needs_supplement` variable, that we can use to determine if the supplement is needed

nonprofittechy commented on docassemble/ALAffidavitOfIndigency/data/questions/affidavit_body.yml:
```diff
@@ -58,6 +57,16 @@ code: |
+  - public_assistance_kinds
+  - household_income_qualifies
+  - can_afford
+code: |
+  need_supplement = not public_assistance_kinds.any_true() and not household_income_qualifies and not can_afford
```
This is a good change--should it also have a `depends on` modifier?

nonprofittechy commented on docassemble/ALAffidavitOfIndigency/data/questions/affidavit_body.yml:
```diff
@@ -58,6 +57,16 @@ code: |
+  - can_afford
+code: |
+  need_supplement = not public_assistance_kinds.any_true() and not household_income_qualifies and not can_afford
+---
+code: |
+  need_income = not public_assistance_kinds.any_true()
```
Similarly, this should get a `depends on` modifier (make sure that works--might need to use an `on change` block instead)

BryceStevenWilley replied:
Can't actually get it work with either. :/  I think it has something to do with the issues I mentioned in https://github.com/SuffolkLITLab/docassemble-ALAffidavitOfIndigency/pull/6#discussion_r982789574.

nonprofittechy reviewed (changes requested):
A few places we added intermediate variables but there's nothing to trigger updates--seems worth adding those so people can review and correct their income and benefits.

We need to make a decision about what to do with all of the disclaimers and other methods we came up with to make sure the email-filed fee waivers had enough information for the clerk. The change to `docket_numbers` breaks that, but maybe we no longer need it. We should at least document and make an intentional choice there. Not sure it needs a full team discussion, but @colarusso may want to weigh in.

The change to subtracting 1 from the household member makes it correct with the current MAPovertyScale repo, but I think the way you fixed it in the newer PovertyScale repo is a lot clearer. The `household_size()` function is also a little confusing--we don't have many functions like that in other places.

Can leave it up to you how to proceed, but not sure we can merge this as-is.

nonprofittechy commented on docassemble/ALAffidavitOfIndigency/data/questions/affidavit_body.yml:
```diff
@@ -169,19 +178,20 @@ buttons:
 ---
 code: |
-  household_monthly_income = hh_income.amount(period_to_use=12)
+  def household_size():
+    if public_assistance_kinds.any_true():
+      return DAEmpty()
+    return household_additional_size + 1 if has_household_members else 1
```
This is pretty confusing

BryceStevenWilley replied:
I removed it for now, but I've been running into a lot of limits of `depends on` and `on change` and making variables into functions for simple calculations seems to be the best:
* `depends on` and `on change` also don't play well with `skip undefined: True`, just like `reconsider: True` does. The attachment block will just output blanks for variables that were undefined
* `depends on` and `on change` also don't seem to play well with lists and gathered attributes changing: variables like `users` don't properly trigger code blocks, I'd guess because DA is just checking variable references, and not all of their elements and attributes. There isn't a reliable way to recalculate those variables: you'd have to list all of the attributes that could possibly cause a variable to change. For example, `case_name` would need to explicitly list:   
    * `users[0].name.first`
    * `users[0].name.middle`
    * `users[0].name.last`
    * `users[1].name.first`
    * ... and so on

nonprofittechy replied:
The dual responsibility of the function was part of the problem--I can remember why I made the value a `DAEmpty` as a quick fix, but it's weird to encapsulate that behavior. Probably best to put that logic in the interview order block and attachment block instead, or rely on `skip undefined: True` at a minimum, but I originally went for a quick fix without knowing all possible solutions (this was originally built in the first weeks of lockdown when getting a remote signature on a fee waiver was important for e-filing, because e-filing also had a per envelope charge).

BryceStevenWilley replied:
Gotcha. I think it's less confusing now, it's still necessary, just because there are quite a few screens that use it. Can't quite simplify as much as I want.

BryceStevenWilley commented on docassemble/ALAffidavitOfIndigency/data/questions/affidavit_body.yml:
```diff
@@ -649,6 +658,7 @@ fields:
       is: False
   - What case do you want fees waived for? Briefly explain how the clerk can find your case.: docket_numbers[0]
     show if: dont_know_docket_number
+    required: False
```
The variable name is fine to me, it was more that the explanation on the page said if it was a new case, to leave the information blank, which isn't possible. I changed that wording, made this required again, and also gave some helpful defaults if someone already said it's a new case.

BryceStevenWilley commented:
I want to test a bit more (probably add some Kiln tests, given how widespread this form is), but I think I've addressed most of the issues. The larger question of review correctness still stands, and I can't really see a way to fix things, DA at the moment just doesn't work well for recalculations.

nonprofittechy commented on docassemble/ALAffidavitOfIndigency/data/questions/affidavit_body.yml:
```diff
@@ -643,21 +668,23 @@ subquestion: |
 fields:
   - I do not know the docket number: dont_know_docket_number
     datatype: yesno
+    default: ${ "Yes" if case_new_or_existing == "new" else "" }
```
I guess this works because of YAML semantics, but could it just be `default: ${ case_new_or_existing == "new" }`?

nonprofittechy replied:
```suggestion
    default: ${ case_new_or_existing == "new" }
```

nonprofittechy reviewed (approved):
Looks good. We'll want to address the `docket_number` issue when we tackle #8 further, but I think this revision works well for now.
